ligo/src/proposal.mligo

```diff
@@ -4,21 +4,361 @@
+#include "management.mligo"
+#include "permit.mligo"
 
-let propose (param, store : propose_params * storage) : return =
-  not_implemented("propose")
+// -----------------------------------------------------------------
+// Helper
+// -----------------------------------------------------------------
 
-let drop_proposal (param, store : proposal_key * storage) : return =
-  not_implemented("drop_proposal")
+[@inline]
+let to_proposal_key (propose_params, sender_addr : propose_params * address): proposal_key =
+  Crypto.blake2b (Bytes.pack (propose_params, sender_addr))
 
-let vote(param, store : vote_param_permited list * storage) : return =
-  not_implemented("vote")
+[@inline]
+let check_if_proposal_exist (proposal_key, store : proposal_key * storage): proposal =
+  match Map.find_opt proposal_key store.proposals with
+    Some p -> p
+  | None -> (failwith("PROPOSAL_NOT_EXIST") : proposal)
 
-let set_voting_period(param, store : voting_period * storage) : return =
-  not_implemented("set_voting_period")
+// TODO: make all checks return storage
 
-let set_quorum_threshold(param, store : voting_period * storage) : return =
-  not_implemented("set_quorum_threshold")
+[@inline]
+let ensure_voting_period_is_not_over (proposal, store : proposal * storage): unit =
+  if Tezos.now >= proposal.start_date + int(store.voting_period)
+    then failwith("VOTING_PERIOD_OVER")
+    else ()
+
+[@inline]
+let ensure_proposal_is_unique (propose_params, store : propose_params * storage): unit =
+  let proposal_key = to_proposal_key(propose_params, sender) in
+  if Map.mem proposal_key store.proposals
+    then failwith("PROPOSAL_NOT_UNIQUE")
+    else ()
+
+// -----------------------------------------------------------------
+// Propose
+// -----------------------------------------------------------------
+
+[@inline]
+let check_is_proposal_valid (config, propose_params, store : config * propose_params * storage): unit =
+  if config.proposal_check (propose_params, store)
+    then ()
+    else failwith("FAIL_PROPOSAL_CHECK")
+
+// TODO: proper balance error
+[@inline]
+let check_proposer_unfrozen_token (propose_params, store : propose_params * storage): unit =
+  let current_balance =
+    match Map.find_opt (Tezos.sender, unfrozen_token_id) store.ledger with
+      None -> (failwith("FA2_INSUFFICIENT_BALANCE") : ledger_value)
+    | Some v -> v
+    in
+  if propose_params.frozen_token > current_balance
+    then failwith("PROPOSAL_INSUFFICIENT_BALANCE")
+    else ()
+
+[@inline]
+let check_proposal_limit_reached (config, propose_params, store : config * propose_params * storage): unit =
+  if config.max_proposals <= List.length store.proposal_key_list_sort_by_date
+    then failwith("MAX_PROPOSALS_REACHED")
+    else ()
+
+[@inline]
+let freeze (tokens, addr, store : nat * address * storage): storage =
+  let store = debit_from (tokens, addr, unfrozen_token_id, store) in
+  let store = credit_to (tokens, addr, frozen_token_id, store) in
+  store
```

Comment:
My version seems a bit better to me, because this way it is easier to visually compare those two lines (with `credit_to` and `debit_from`) and spot a bug if it is present.

Reply:
Fair enough, you have a point there :+1: